src/target/codegen.cc

```diff
@@ -110,32 +115,100 @@ class ModuleSerializer {
           visited.insert(next);
           stack.push_back(next);
         }
       }
+    };
+
+    std::vector<runtime::ModuleNode*> dso_exportable_boundary;
+
+    // Create module index that merges all dso module into a single group.
+    //
+    // Do a two phase visit, to ensure dso module's index
+    // is always bigger than a parent of any dso module
+    // and smaller than children of any dso module.
+    //
+    // Error will be raised in CreateImportTree
+    // if merging dso module causes a cycle in the import tree
+
+    // Phase 0: only expand non-dso-module and record the boundary.
+    while (!stack.empty()) {
+      runtime::ModuleNode* n = stack.back();
+      stack.pop_back();
+      if (DSOExportable(n)) {
+        // do not recursively expand dso modules
+        // we will expand in phase 1
+        dso_exportable_boundary.emplace_back(n);
+      } else {
+        // expand the non-dso modules
+        mod2index_[n] = module_index++;
+        mod_group_vec_.emplace_back(std::vector<runtime::ModuleNode*>({n}));
+        fpush_imports_to_stack(n);
+      }
+    }
+    if (dso_exportable_boundary.size() == 0) return;
+
+    // create the slot for dso exportable modules
+    uint64_t dso_module_index = module_index++;
+    mod_group_vec_.emplace_back(std::vector<runtime::ModuleNode*>());
+
+    stack = std::move(dso_exportable_boundary);
+
+    // Phase 1: expand the children of dso modules.
+    while (!stack.empty()) {
+      runtime::ModuleNode* n = stack.back();
+      stack.pop_back();
+
+      if (DSOExportable(n)) {
+        mod_group_vec_[dso_module_index].emplace_back(n);
+        mod2index_[n] = dso_module_index;
+      } else {
+        mod2index_[n] = module_index++;
+        mod_group_vec_.emplace_back(std::vector<runtime::ModuleNode*>({n}));
+      }
+      fpush_imports_to_stack(n);
     }
   }
 
   void CreateImportTree() {
-    for (auto m : mod_vec_) {
-      for (runtime::Module im : m->imports()) {
-        uint64_t mod_index = mod2index_[im.operator->()];
-        import_tree_child_indices_.push_back(mod_index);
+    std::vector<int64_t> child_indices;
+
+    for (size_t parent_index = 0; parent_index < mod_group_vec_.size(); ++parent_index) {
+      child_indices.clear();
+      for (const auto* m : mod_group_vec_[parent_index]) {
+        for (runtime::Module im : m->imports()) {
+          uint64_t mod_index = mod2index_.at(im.operator->());
+          // skip cycle when dso modules are merged together
+          if (mod_index != parent_index) {
+            child_indices.emplace_back(mod_index);
+          }
+        }
+      }
+      // sort and unique the merged indices
+      std::sort(child_indices.begin(), child_indices.end());
+      auto unique_end = std::unique(child_indices.begin(), child_indices.end());
+
+      // Check cycles due to merging dso exportable modules.
+      if (child_indices.size() != 0) {
+        CHECK_LT(parent_index, child_indices[0])
```

Comment:
it's clear if you read the code carefully, but perhaps to help people, would be great if you could:
1) add a comment here explaining explicitly why lower indices are "higher" in the tree (just explain the traversal order)
2) include the offending module instances in the error message, at least for now this will allow their type_key to be shown to the user and it may help them understand the problem in this case or us understand on forum when there are bug reports or questions.